File: sources/Android/android-communications/library/src/test/java/com/polar/androidcommunications/common/ble/TypeUtilsTest.kt
package com.polar.androidcommunications.common.ble

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File
import java.io.FileReader

class TypeUtilsTest {

    @Test
    fun `test array conversion to unsigned byte max value`() {
        // Arrange
        val byteArray = byteArrayOf(0xFF.toByte())
        val expectedValue = 0xFFu.toUByte()

        // Act
        val result = TypeUtils.convertArrayToUnsignedByte(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to unsigned int max value`() {
        // Arrange
        val byteArray = byteArrayOf(0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte())
        val expectedValue = 0xFFFFFFFFu

        // Act
        val result = TypeUtils.convertArrayToUnsignedInt(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to unsigned int min value`() {
        // Arrange
        val byteArray = byteArrayOf(0x00, 0x00, 0x00, 0x00)
        val expectedValue = 0x00000000u

        // Act
        val result = TypeUtils.convertArrayToUnsignedInt(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to unsigned int max positive int`() {
        // Arrange
        val byteArray = byteArrayOf(0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0x7F.toByte())
        val expectedValue = 0x7FFFFFFFu

        // Act
        val result = TypeUtils.convertArrayToUnsignedInt(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to unsigned int small array`() {
        // Arrange
        val byteArray = byteArrayOf(0xFF.toByte(), 0xFF.toByte())
        val expectedValue = 0xFFFFu

        // Act
        val result = TypeUtils.convertArrayToUnsignedInt(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to unsigned int too big array`() {
        // Arrange
        val byteArray = byteArrayOf(0x00, 0x00, 0x00, 0x00, 0x00)

        // Act & Assert
        assertThrows(AssertionError::class.java) {
            TypeUtils.convertArrayToUnsignedInt(byteArray)
        }
    }

    @Test
    fun `test array conversion to unsigned long max value`() {
        // Arrange
        val byteArray = byteArrayOf(0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte())
        val expectedValue = 0xFFFFFFFFFFFFFFFFu

        // Act
        val result = TypeUtils.convertArrayToUnsignedLong(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to unsigned long min value`() {
        // Arrange
        val byteArray = byteArrayOf(0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
        val expectedValue = 0x0000000000000000u.toULong()

        // Act
        val result = TypeUtils.convertArrayToUnsignedLong(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to unsigned long max positive int`() {
        // Arrange
        val byteArray = byteArrayOf(0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0x7F.toByte())
        val expectedValue = 0x7FFFFFFFFFFFFFFFu

        // Act
        val result = TypeUtils.convertArrayToUnsignedLong(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to unsigned long small array`() {
        // Arrange
        val byteArray = byteArrayOf(0xFF.toByte(), 0xFF.toByte())
        val expectedValue = 0xFFFFu.toULong()

        // Act
        val result = TypeUtils.convertArrayToUnsignedLong(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to unsigned long too big array`() {
        // Arrange
        val byteArray = byteArrayOf(0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00)

        // Act & Assert
        assertThrows(AssertionError::class.java) {
            TypeUtils.convertArrayToUnsignedLong(byteArray)
        }
    }

    @Test
    fun `test array conversion to signed int max value`() {
        // Arrange
        val byteArray = byteArrayOf(0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte())
        val expectedValue = -1

        // Act
        val result = TypeUtils.convertArrayToSignedInt(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to signed int min value`() {
        // Arrange
        val byteArray = byteArrayOf(0x00, 0x00, 0x00, 0x00)
        val expectedValue = 0

        // Act
        val result = TypeUtils.convertArrayToSignedInt(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to signed int max positive int`() {
        // Arrange
        val byteArray = byteArrayOf(0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0x7F.toByte())
        val expectedValue = Int.MAX_VALUE

        // Act
        val result = TypeUtils.convertArrayToSignedInt(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }

    @Test
    fun `test array conversion to signed int small array`() {
        // Arrange
        val byteArray = byteArrayOf(0xFF.toByte())
        val expectedValue = -1

        // Act
        val result = TypeUtils.convertArrayToSignedInt(byteArray)

        // Assert
        assertEquals(expectedValue, result)
    }


    @Test
    fun `test array conversion to signed int too big array`() {
        // Arrange
        val byteArray = byteArrayOf(0x00, 0x00, 0x00, 0x00, 0x00)

        // Act & Assert
        assertThrows(AssertionError::class.java) {
            TypeUtils.convertArrayToSignedInt(byteArray)
        }
    }

    @Test
    fun `test conversion unsigned byte to int`() {
        // Arrange
        val testByte1: Byte = 0x00.toByte()
        val testByte2: Byte = 0x80.toByte()
        val testByte3: Byte = 0xFF.toByte()
        val testByte4: Byte = 0x55.toByte()

        // Act & Assert
        assertEquals(0, TypeUtils.convertUnsignedByteToInt(testByte1))
        assertEquals(128, TypeUtils.convertUnsignedByteToInt(testByte2))
        assertEquals(255, TypeUtils.convertUnsignedByteToInt(testByte3))
        assertEquals(85, TypeUtils.convertUnsignedByteToInt(testByte4))

    }

    @Test
    fun typeUtilsGoldenVectors_matchAndroidBehavior() {
        val vectors = loadTypeUtilsVectors()
        assertTrue("Expected type utility golden vectors", vectors.isNotEmpty())

        vectors.forEach { vector ->
            val input = vector.getAsJsonObject("input")
            val expected = vector.platformExpected("android")
            val bytes = input.get("hex").asString.hexToByteArray()
            val caseId = vector.get("id").asString
            val offset = input.get("offset")?.asInt
            val size = input.get("size")?.asInt

            if (expected.has("unsignedByte")) {
                assertEquals(caseId, expected.get("unsignedByte").asInt.toUByte(), TypeUtils.convertArrayToUnsignedByte(bytes))
            }

            if (expected.has("unsignedInt")) {
                assertEquals(caseId, expected.get("unsignedInt").asLong.toUInt(), bytes.convertToUnsignedInt(offset, size))
            }

            if (expected.has("unsignedLong")) {
                assertEquals(caseId, expected.get("unsignedLong").asString.toULong(), bytes.convertToUnsignedLong(offset, size))
            }

            if (expected.has("signedInt")) {
                assertEquals(caseId, expected.get("signedInt").asInt, bytes.convertToSignedInt(offset, size))
            }

            if (expected.has("unsignedIntError")) {
                assertThrows(caseId, AssertionError::class.java) {
                    bytes.convertToUnsignedInt(offset, size)
                }
            }

            if (expected.has("unsignedLongError")) {
                assertThrows(caseId, AssertionError::class.java) {
                    bytes.convertToUnsignedLong(offset, size)
                }
            }

            if (expected.has("signedIntError")) {
                assertThrows(caseId, AssertionError::class.java) {
                    bytes.convertToSignedInt(offset, size)
                }
            }
        }
    }

    @Test
    fun `type utility golden vectors follow neutral KMP vector shape`() {
        loadTypeUtilsVectors().forEach { vector ->
            val id = vector.get("id")?.asString ?: "unknown-vector"

            assertTrue(id, vector.has("area"))
            assertTrue(id, vector.has("case"))
            assertTrue(id, vector.has("source"))
            assertTrue(id, vector.has("input"))
            assertTrue(id, vector.has("expected"))
            assertTrue(id, vector.has("platforms"))
            val platforms = vector.getAsJsonObject("platforms")
            assertTrue(id, platforms.has("android"))
            assertTrue(id, platforms.has("ios"))
            assertTrue(id, platforms.has("common"))
        }
    }

    @Test
    fun `type utils readiness manifest is pinned before parser primitive migration`() {
        val readiness = loadTypeUtilsReadinessManifest()
        val input = readiness.getAsJsonObject("input")
        val expected = readiness.getAsJsonObject("expected")
        val consumerTests = readiness.getAsJsonObject("consumerTests")
        val policyVectorPaths = input.getAsJsonArray("policyVectorPaths").map { it.asString }
        val requiredFamilies = input.getAsJsonArray("requiredBehaviorFamilies").map { it.asString }
        val coveredFamilies = expected.getAsJsonArray("coveredBehaviorFamilies").map { it.asString }

        assertEquals("type-utils-readiness", readiness.get("id").asString)
        assertEquals("typeUtilsReadiness", input.get("kind").asString)
        assertEquals(
            listOf(
                "protocol/type-utils/empty-payload-platform-difference.json",
                "protocol/type-utils/offset-signed-int-negative-boundary.json",
                "protocol/type-utils/offset-unsigned-int-little-endian.json",
                "protocol/type-utils/signed-int-24bit-negative-one.json",
                "protocol/type-utils/signed-int-max.json",
                "protocol/type-utils/signed-int-min-16bit.json",
                "protocol/type-utils/signed-int-min-24bit.json",
                "protocol/type-utils/signed-int-min-32bit.json",
                "protocol/type-utils/signed-int-negative-one.json",
                "protocol/type-utils/signed-int-too-long.json",
                "protocol/type-utils/unsigned-byte-max.json",
                "protocol/type-utils/unsigned-int-high-bit-16bit-platform-difference.json",
                "protocol/type-utils/unsigned-int-high-bit-platform-difference.json",
                "protocol/type-utils/unsigned-int-little-endian.json",
                "protocol/type-utils/unsigned-int-too-long.json",
                "protocol/type-utils/unsigned-long-max.json",
                "protocol/type-utils/unsigned-long-too-long.json"
            ),
            policyVectorPaths
        )
        val expectedFamilies = listOf(
            "unsigned-byte-conversion",
            "little-endian-unsigned-int-conversion",
            "little-endian-unsigned-long-conversion",
            "signed-int-sign-extension",
            "offset-and-size-selection",
            "signed-minimum-boundaries",
            "unsigned-high-bit-platform-decision",
            "empty-payload-error-policy",
            "payload-too-long-error-policy",
            "uint64-max-decimal-preservation",
            "platform-type-utils-vector-reference-gate",
            "compile-verification-gate"
        )
        assertEquals(expectedFamilies, requiredFamilies)
        assertEquals(expectedFamilies, coveredFamilies)
        assertEquals(
            "Type utility migration may proceed only after every vector named by this readiness manifest is executable from shared commonTest, Android and iOS type utility tests continue to reference the same vectors, unsigned byte/int/long conversion, signed sign extension, offset and size selection, signed-minimum boundaries, high-bit unsigned platform decisions, empty payload and payload-too-long typed errors, UInt64 max decimal preservation, and compile verification remain explicit before production parser primitives move.",
            expected.get("commonDecision").asString
        )
        assertEquals(listOf("com.polar.androidcommunications.common.ble.TypeUtilsTest"), consumerTests.getAsJsonArray("android").map { it.asString })
        assertEquals(listOf("TypeUtilsTest"), consumerTests.getAsJsonArray("ios").map { it.asString })
        assertEquals(listOf("com.polar.sharedtest.TypeUtilsCommonPolicyTest"), consumerTests.getAsJsonArray("commonPrototype").map { it.asString })
    }

    private fun JsonObject.platformExpected(platform: String): JsonObject {
        return if (has("platformExpectations")) {
            getAsJsonObject("platformExpectations").getAsJsonObject(platform)
        } else {
            getAsJsonObject("expected")
        }
    }

    private fun ByteArray.convertToUnsignedInt(offset: Int?, size: Int?): UInt {
        return if (offset != null && size != null) {
            TypeUtils.convertArrayToUnsignedInt(this, offset, size)
        } else {
            TypeUtils.convertArrayToUnsignedInt(this)
        }
    }

    private fun ByteArray.convertToUnsignedLong(offset: Int?, size: Int?): ULong {
        return if (offset != null && size != null) {
            TypeUtils.convertArrayToUnsignedLong(this, offset, size)
        } else {
            TypeUtils.convertArrayToUnsignedLong(this)
        }
    }

    private fun ByteArray.convertToSignedInt(offset: Int?, size: Int?): Int {
        return if (offset != null && size != null) {
            TypeUtils.convertArrayToSignedInt(this, offset, size)
        } else {
            TypeUtils.convertArrayToSignedInt(this)
        }
    }

    private fun loadTypeUtilsVectors(): List<JsonObject> {
        val vectorDirectory = findRepositoryRoot()
            .resolve("testdata/golden-vectors/protocol/type-utils")
        return vectorDirectory
            .listFiles { file -> file.isFile && file.extension == "json" }
            .orEmpty()
            .sortedBy { it.name }
            .map { file ->
                FileReader(file).use { reader ->
                    JsonParser().parse(reader).asJsonObject
                }
            }
            .filterNot { vector -> vector.getAsJsonObject("input")?.get("kind")?.asString == "typeUtilsReadiness" }
    }

    private fun loadTypeUtilsReadinessManifest(): JsonObject {
        val vectorFile = findRepositoryRoot()
            .resolve("testdata/golden-vectors/protocol/type-utils/type-utils-readiness.json")
        FileReader(vectorFile).use { reader ->
            return JsonParser().parse(reader).asJsonObject
        }
    }

    private fun findRepositoryRoot(): File {
        val userDirectory = System.getProperty("user.dir") ?: error("user.dir is not set")
        var directory = File(userDirectory).absoluteFile
        while (true) {
            if (directory.resolve("testdata/golden-vectors").isDirectory) {
                return directory
            }
            directory = directory.parentFile ?: error("Could not find repository root from $userDirectory")
        }
    }

    private fun String.hexToByteArray(): ByteArray {
        require(length % 2 == 0) { "Hex string must have an even length" }
        return chunked(2)
            .map { it.toInt(16).toByte() }
            .toByteArray()
    }
}

File: sources/iOS/ios-communications/Tests/iOSCommunicationsTests/TypeUtilsTest.swift
//  Copyright © 2022 Polar. All rights reserved.

import XCTest
@testable import iOSCommunications

final class TypeUtilsTest: XCTestCase {

    override func setUpWithError() throws {
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDownWithError() throws {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }

    func testArrayConversionToSignedIntMaxValue() throws {
        // Arrange
        let byteArray = Data([0xFF, 0xFF, 0xFF, 0xFF])
        let expectedValue:Int32 = -1

        // Act
        let result = TypeUtils.convertArrayToSignedInt(byteArray)

        // Assert
        XCTAssertEqual(expectedValue, result)
    }
    
    func testArrayConversionToSignedIntMinValue() throws {
        // Arrange
        let byteArray = Data([0x00, 0x00, 0x00, 0x00])
        let expectedValue:Int32 = 0

        // Act
        let result = TypeUtils.convertArrayToSignedInt(byteArray)

        // Assert
        XCTAssertEqual(expectedValue, result)
    }
    
    func testArrayConversionToSignedIntMaxPositiveInt() throws {
        // Arrange
        let byteArray = Data([0xFF, 0xFF, 0xFF, 0x7F])
        let expectedValue = Int32.max

        // Act
        let result = TypeUtils.convertArrayToSignedInt(byteArray)

        // Assert
        XCTAssertEqual(expectedValue, result)
    }

    func testArrayConversionToSignedIntSmallArray() throws {
        // Arrange
        let byteArray = Data([0xFF])
        let expectedValue:Int32 = -1

        // Act
        let result = TypeUtils.convertArrayToSignedInt(byteArray)

        // Assert
        XCTAssertEqual(expectedValue, result)
    }

    func testTypeUtilsUsesProductionConversionPolicy() throws {
        XCTAssertEqual(Int32(-32768), TypeUtils.convertArrayToSignedInt(Data([0x00, 0x80])))
        XCTAssertEqual(Int32(-32768), TypeUtils.convertArrayToSignedInt(Data([0x7f, 0x00, 0x80, 0x01]), offset: 1, size: 2))
        XCTAssertEqual(UInt64.max, TypeUtils.convertArrayToUnsignedInt64(Data([0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff])))
        XCTAssertEqual(UInt64.max, TypeUtils.convertArrayToUnsignedInt64(Data([0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff]), offset: 1, size: 8))
    }

    func testTypeUtilsGoldenVectorsMatchIOSCommunicationsBehavior() throws {
        let vectors = try loadTypeUtilsGoldenVectors()
        XCTAssertFalse(vectors.isEmpty, "Expected type utility golden vectors")

        for vector in vectors {
            let id = vector["id"] as? String ?? "unknown-vector"
            let input = try XCTUnwrap(vector["input"] as? [String: Any], id)
            let expected = try platformExpected(vector, platform: "ios")
            let hex = try XCTUnwrap(input["hex"] as? String, id)
            let data = try Data(hexString: hex)
            let offset = input["offset"] as? Int
            let size = input["size"] as? Int

            if let expectedUnsignedInt = expected["unsignedInt"] as? NSNumber {
                XCTAssertEqual(UInt(expectedUnsignedInt.uint64Value), convertArrayToUnsignedInt(data, offset: offset, size: size), id)
            } else if let expectedUnsignedInt = expected["unsignedInt"] as? String {
                XCTAssertEqual(expectedUnsignedInt, String(convertArrayToUnsignedInt(data, offset: offset, size: size)), id)
            }

            if let expectedUnsignedLong = expected["unsignedLong"] as? NSNumber {
                XCTAssertEqual(UInt64(expectedUnsignedLong.uint64Value), convertArrayToUnsignedInt64(data, offset: offset, size: size), id)
            } else if let expectedUnsignedLong = expected["unsignedLong"] as? String {
                XCTAssertEqual(UInt64(expectedUnsignedLong), convertArrayToUnsignedInt64(data, offset: offset, size: size), id)
            }

            if let expectedSignedInt = expected["signedInt"] as? NSNumber {
                XCTAssertEqual(expectedSignedInt.int32Value, convertArrayToSignedInt(data, offset: offset, size: size), id)
            }

            if expected["unsignedIntError"] != nil {
                assertConversionErrorPrecondition(data: data, expected: expected["unsignedIntError"], maxSize: 4, id: id)
            }

            if expected["unsignedLongError"] != nil {
                assertConversionErrorPrecondition(data: data, expected: expected["unsignedLongError"], maxSize: 8, id: id)
            }

            if expected["signedIntError"] != nil {
                assertConversionErrorPrecondition(data: data, expected: expected["signedIntError"], maxSize: 4, id: id)
            }
        }
    }

    func testTypeUtilsGoldenVectorsFollowNeutralKmpVectorShape() throws {
        for vector in try loadTypeUtilsGoldenVectors() {
            let id = vector["id"] as? String ?? "unknown-vector"

            XCTAssertNotNil(vector["area"], id)
            XCTAssertNotNil(vector["case"], id)
            XCTAssertNotNil(vector["source"], id)
            XCTAssertNotNil(vector["input"], id)
            XCTAssertNotNil(vector["expected"], id)
            let platforms = try XCTUnwrap(vector["platforms"] as? [String: Any], id)
            XCTAssertNotNil(platforms["android"], id)
            XCTAssertNotNil(platforms["ios"], id)
            XCTAssertNotNil(platforms["common"], id)
        }
    }

    func testTypeUtilsReadinessManifestIsPinnedBeforeParserPrimitiveMigration() throws {
        let readiness = try loadTypeUtilsReadinessManifest()
        let input = try XCTUnwrap(readiness["input"] as? [String: Any])
        let expected = try XCTUnwrap(readiness["expected"] as? [String: Any])
        let consumerTests = try XCTUnwrap(readiness["consumerTests"] as? [String: Any])
        let platforms = try XCTUnwrap(readiness["platforms"] as? [String: Any])
        let policyVectorPaths = try XCTUnwrap(input["policyVectorPaths"] as? [String])
        let requiredFamilies = try XCTUnwrap(input["requiredBehaviorFamilies"] as? [String])
        let coveredFamilies = try XCTUnwrap(expected["coveredBehaviorFamilies"] as? [String])
        XCTAssertEqual(readiness["id"] as? String, "type-utils-readiness")
        XCTAssertEqual(input["kind"] as? String, "typeUtilsReadiness")
        XCTAssertEqual(policyVectorPaths, [
            "protocol/type-utils/empty-payload-platform-difference.json",
            "protocol/type-utils/offset-signed-int-negative-boundary.json",
            "protocol/type-utils/offset-unsigned-int-little-endian.json",
            "protocol/type-utils/signed-int-24bit-negative-one.json",
            "protocol/type-utils/signed-int-max.json",
            "protocol/type-utils/signed-int-min-16bit.json",
            "protocol/type-utils/signed-int-min-24bit.json",
            "protocol/type-utils/signed-int-min-32bit.json",
            "protocol/type-utils/signed-int-negative-one.json",
            "protocol/type-utils/signed-int-too-long.json",
            "protocol/type-utils/unsigned-byte-max.json",
            "protocol/type-utils/unsigned-int-high-bit-16bit-platform-difference.json",
            "protocol/type-utils/unsigned-int-high-bit-platform-difference.json",
            "protocol/type-utils/unsigned-int-little-endian.json",
            "protocol/type-utils/unsigned-int-too-long.json",
            "protocol/type-utils/unsigned-long-max.json",
            "protocol/type-utils/unsigned-long-too-long.json"
        ])
        let expectedFamilies = [
            "unsigned-byte-conversion",
            "little-endian-unsigned-int-conversion",
            "little-endian-unsigned-long-conversion",
            "signed-int-sign-extension",
            "offset-and-size-selection",
            "signed-minimum-boundaries",
            "unsigned-high-bit-platform-decision",
            "empty-payload-error-policy",
            "payload-too-long-error-policy",
            "uint64-max-decimal-preservation",
            "platform-type-utils-vector-reference-gate",
            "compile-verification-gate"
        ]
        XCTAssertEqual(requiredFamilies, expectedFamilies)
        XCTAssertEqual(coveredFamilies, expectedFamilies)
        XCTAssertEqual(expected["commonDecision"] as? String, "Type utility migration may proceed only after every vector named by this readiness manifest is executable from shared commonTest, Android and iOS type utility tests continue to reference the same vectors, unsigned byte/int/long conversion, signed sign extension, offset and size selection, signed-minimum boundaries, high-bit unsigned platform decisions, empty payload and payload-too-long typed errors, UInt64 max decimal preservation, and compile verification remain explicit before production parser primitives move.")
        XCTAssertEqual(try XCTUnwrap(consumerTests["android"] as? [String]), ["com.polar.androidcommunications.common.ble.TypeUtilsTest"])
        XCTAssertEqual(try XCTUnwrap(consumerTests["ios"] as? [String]), ["TypeUtilsTest"])
        XCTAssertEqual(try XCTUnwrap(consumerTests["commonPrototype"] as? [String]), ["com.polar.sharedtest.TypeUtilsCommonPolicyTest"])
        XCTAssertEqual(platforms["android"] as? Bool, true)
        XCTAssertEqual(platforms["ios"] as? Bool, true)
        XCTAssertEqual(platforms["common"] as? Bool, true)
    }

    private func loadTypeUtilsGoldenVectors() throws -> [[String: Any]] {
        let vectorDirectory = try GoldenVectorTestData.repositoryRoot()
            .appendingPathComponent("testdata/golden-vectors/protocol/type-utils")
        return try FileManager.default
            .contentsOfDirectory(at: vectorDirectory, includingPropertiesForKeys: nil)
            .filter { $0.pathExtension == "json" }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }
            .map { file in
                let data = try Data(contentsOf: file)
                return try XCTUnwrap(JSONSerialization.jsonObject(with: data) as? [String: Any], file.path)
            }
            .filter { vector in
                let input = vector["input"] as? [String: Any]
                return input?["kind"] as? String != "typeUtilsReadiness"
            }
    }

    private func loadTypeUtilsReadinessManifest() throws -> [String: Any] {
        let vectorFile = try GoldenVectorTestData.repositoryRoot()
            .appendingPathComponent("testdata/golden-vectors/protocol/type-utils/type-utils-readiness.json")
        let data = try Data(contentsOf: vectorFile)
        return try XCTUnwrap(JSONSerialization.jsonObject(with: data) as? [String: Any], vectorFile.path)
    }


    private func platformExpected(_ vector: [String: Any], platform: String) throws -> [String: Any] {
        if let platformExpectations = vector["platformExpectations"] as? [String: Any] {
            return try XCTUnwrap(platformExpectations[platform] as? [String: Any], vector["id"] as? String ?? "unknown-vector")
        }
        return try XCTUnwrap(vector["expected"] as? [String: Any], vector["id"] as? String ?? "unknown-vector")
    }

    private func convertArrayToUnsignedInt(_ data: Data, offset: Int?, size: Int?) -> UInt {
        if let offset, let size {
            return TypeUtils.convertArrayToUnsignedInt(data, offset: offset, size: size)
        }
        return TypeUtils.convertArrayToUnsignedInt(data)
    }

    private func convertArrayToUnsignedInt64(_ data: Data, offset: Int?, size: Int?) -> UInt64 {
        if let offset, let size {
            return TypeUtils.convertArrayToUnsignedInt64(data, offset: offset, size: size)
        }
        return TypeUtils.convertArrayToUnsignedInt64(data)
    }

    private func convertArrayToSignedInt(_ data: Data, offset: Int?, size: Int?) -> Int32 {
        if let offset, let size {
            return TypeUtils.convertArrayToSignedInt(data, offset: offset, size: size)
        }
        return TypeUtils.convertArrayToSignedInt(data)
    }

    private func assertConversionErrorPrecondition(data: Data, expected: Any?, maxSize: Int, id: String) {
        switch expected as? String {
        case "payloadTooLong":
            XCTAssertGreaterThan(data.count, maxSize, id)
        case "emptyPayloadPrecondition":
            XCTAssertTrue(data.isEmpty, id)
        case "emptyPayload":
            XCTAssertTrue(data.isEmpty, id)
        default:
            XCTFail("Unsupported conversion error expectation in \(id): \(String(describing: expected))")
        }
    }
}

private extension Data {
    init(hexString: String) throws {
        guard hexString.count.isMultiple(of: 2) else {
            throw NSError(domain: "TypeUtilsTest", code: 2, userInfo: [NSLocalizedDescriptionKey: "Hex string must have an even length"])
        }
        var bytes: [UInt8] = []
        var index = hexString.startIndex
        while index < hexString.endIndex {
            let nextIndex = hexString.index(index, offsetBy: 2)
            let byteString = String(hexString[index..<nextIndex])
            guard let byte = UInt8(byteString, radix: 16) else {
                throw NSError(domain: "TypeUtilsTest", code: 3, userInfo: [NSLocalizedDescriptionKey: "Invalid hex byte \(byteString)"])
            }
            bytes.append(byte)
            index = nextIndex
        }
        self.init(bytes)
    }
}
